src/libsyntax/parse/parser.rs

```diff
@@ -1753,21 +1753,65 @@ impl<'a> Parser<'a> {
             (pat, self.parse_ty()?)
         } else {
             debug!("parse_arg_general ident_to_pat");
-            let ident = Ident::new(keywords::Invalid.name(), self.prev_span);
-            let ty = self.parse_ty()?;
-            let pat = P(Pat {
-                id: ast::DUMMY_NODE_ID,
-                node: PatKind::Ident(BindingMode::ByValue(Mutability::Immutable), ident, None),
-                span: ty.span,
-            });
-            (pat, ty)
+
+            // If we see `ident :`, then we know that the argument is not just of the
+            // form `type`, which means we won't need to recover from parsing a
+            // pattern and so we don't need to store a parser snapshot.
+            let parser_snapshot_before_pat = if
+                self.look_ahead(1, |t| t.is_ident()) &&
+                self.look_ahead(2, |t| t == &token::Colon) {
```

Comment:
This would still store a copy of the parser for `fn foo(&mut x) {}` and `fn bar(&self)`, right?

Reply:
Yes.